Diesel tool sounds - First Draft / Feedback desired

MalkContent opened the pull request:
This works well enough, but it's like 13 workarounds in a trench coat.
I have some thoughts about how to do some things differently, but I figured a) perfect is the enemy of good and b) it's probably a good idea to get some feedback on this to see if just what the PR does is in the right vein
technical feedback of course welcome, too ^^

resolves https://github.com/BluSunrize/ImmersiveEngineering/issues/3262

Commit:
Initial commit

Commit:
first presentable draft

MalkContent commented:
resolves https://github.com/BluSunrize/ImmersiveEngineering/issues/3262

voidsong-dragonfly commented on src/main/java/blusunrize/immersiveengineering/common/util/sound/DieselToolSoundHandler.java:
~~Why have all of this? Why not do something similar to how `handleTileSound(...)` in `ClientProxy.java` works? That seems like it would both be easier and more effective, though I may be off my rocket and confused here.~~ Once talking with Malk, I think this is probably not reasonable to do.

voidsong-dragonfly reviewed (changes requested):
First thing, Malk: These sounds are extremely loud and close to the player. I would suggest making the sounds maybe a quarter again as quiet to ensure they're not too disruptive to the player experience.
Second thing: The harvest detection you implemented for the drill/busszaw does not work particularly well **between blocks of the same type**. When breaking grass blocks just as a test, the drill seemed to keep puttering up and down and not actually hitting a stride and running. Within one block it worked great.
Third thing: All the drill sounds and the saw sounds except for the saw cutting sound great. However, the saw cutting sounds... very electronic? And might be better for the angle grinder only, and have something else for the sawblade. Maybe the sawmill?

Commit:
pulled out tool sounds from DieselToolItem into INoisyTool interface

Commit:
made get<*>Sound calls take ItemStack as argument
made buzzsaw harvest sound depending on the installed saw blade
gave previous buzzsaw harvest sound to grinder/rockcutter blades
copied saw_full.ogg for the regular saw blade as a temporary placeholder

Commit:
added a temporary static for volume adjustment in INoisyTool

Commit:
added a fading sound to the tools to make digging quickly mined blocks sound less like *checks notes* "cranking a mower" :D

MalkContent commented:
to First: I put in a (hopefully temporary) static for you to control all the sounds volume with. I recon it's easier for you guys to adjust the volume until you find something you're comfortable with, then I can adjust the sounds themselves to that level. My original take was "loud noisy tool, use ear defenders when annoyed"

to Second: after having cleared up what you mean (sound jumping up and down between quickly digged blocks), I added a fade sound after digging, which does sound much nicer imo. I had already mulled over doing that previously, but originally decided against it, because I wanted to KISS it to not make Blu groan harder than I assume this PR does anyways x)

to Third: I agree about the saw. Had a hard time to find a good sound that does not sound "far away with echo, in a factory hall". I hanged it so the sounds depend on the ItemStack, not just the Item when fetched, so now the rockcutter and grinding disk have the grindy sound and the good old saw blade has a copy of the saw mill's saw_full.ogg as a temporary place holder.

-> Experimenting, because I initially did not understand your description in "Second" correctly made me find a bug with a double sound when you start digging under specific circumstances, though. Have not nailed it down yet, though.
And for the saw (blade) sawing sound, I'll have another look around, but that'll probably be a bit later in the weekend.

Commit:
fixed single-stutter bug when beginning to harvest
some refactoring in NoisyToolSoundGroup
corrected sound instance class names from Diesel* to Noisy*

Commit:
added important remarks to the INoisyTool

MalkContent commented:
fixed the bug ^_^
now just gotta see about the sawing sound

Commit:
fixed follow-up stutter bug from fixing the first stutter bug 🤡

Commit:
correct checkItemMatch to compare stacks not items, since sounds are now item & stack dependent (could lead to switching hand item to differently sounding stack of the same item, but keeping the sounds of the previous item)
also makes switching between same-item hand items more distinct

Commit:
fix edge case NPE

Commit:
replace placeholder saw blade sound

Commit:
rebalance buzzsaw grinding sound volume to be more in line with the other NoisyTool sounds

Commit:
fixed checkItemMatch failing because updating any values on a tool item stack (ie. damage/fuel use) ultimatively creates a new ItemStack
now it is 14 workarounds in a trench coat
removed unnecessary public keywords in INoisyTool
removed unnecessary formatter disable

Commit:
made the same-stack check item based instead of static
fix: if the ItemStacks are not identical when checked in checkItemValid, but still considered the same, the ItemStack of the sound group updates

Commit:
clear sound groups on leaving level (change level or back to menu)

voidsong-dragonfly commented:
There's still a bit of the "jumpiness" I mentioned, but you're right that there is a lot less of it and it sounds less annoying. Not sure if there is a good way to fix it, with how the delays between block breaking works. For now, this seems... fine enough? I'll wait for Blu to weigh in on what she thinks about the PR, though.

MalkContent commented:
yea, I can understand that. I'm open to ideas there, but my ideas on "how to do this differently" basically boil down to "extend the digging sound beyond the point where the block breaks to bridge the gap", which probably sounds weird with the block breaking, but the digging sound still going, or "do button press hitscan stuff and check if the player <would> continue digging, if he keeps the button pressed", which could maybe work nicely, but I also am not looking forward to working that into the state machine.

And, jumpiness aside, my argument in favor of the current behaviour is "it does tell you clearly when you're digging and when you're not".

Maybe the sounds themselves could be addressed to smoothe the jumpiness somewhat. But I'm taking a couple days to cleansy my ears' palate from listening to saw sounds for hours x)

Commit:
post rebase cleanup/bugfix and update to 1.21

MalkContent commented:
hoh boy. I'll look at that datagen error with fresh eyes some later date

Commit:
added missing generated resource lang file(?)

Commit:
fixed wrong symbol(?) I don't even wanna know how that came into being

Commit:
added missing datagen-ed .cache file

malte0811 committed:
Rerun datagen

BluSunrize commented on src/main/java/blusunrize/immersiveengineering/common/items/BuzzsawItem.java:
```diff
@@ -20,6 +20,8 @@
 import blusunrize.immersiveengineering.common.register.IEItems.Misc;
 import blusunrize.immersiveengineering.common.register.IEItems.Tools;
+import blusunrize.immersiveengineering.common.util.IESounds;
+//import blusunrize.immersiveengineering.common.util.ItemNBTHelper;
```
We do not do commenting out of imports.

BluSunrize commented on src/main/java/blusunrize/immersiveengineering/common/util/sound/NoisyToolSoundGroup.java:
```diff
@@ -0,0 +1,292 @@
+				if(propagate)
+					updateHarvestState(null, false); // todo: check if this is really necessary. Not that it hurts.
+				play(new NoisyToolMotorSoundFinite(noisyToolItem.getAttackSound(noisyToolStack).value(), newMotorState, ATTACK_DURATION));
+				break;
```
Attack sounds are currently not playing for me?

BluSunrize commented on src/main/resources/assets/immersiveengineering/lang_base/en_us.json:
```diff
@@ -462,6 +462,13 @@
+  "subtitle.immersiveengineering.buzzsaw_motor": "Buzzsaw motor running",
+  "subtitle.immersiveengineering.buzzsaw_attack": "Buzzsaw revs up high",
+  "subtitle.immersiveengineering.buzzsaw_harvest_sawing": "Sawing",
+  "subtitle.immersiveengineering.buzzsaw_harvest_grinding": "Grinding",
```
The buzzsaw harvest subtitles do not match what it listed in the sound file

BluSunrize reviewed (changes requested):
I finally checked this over, and I actually quite like it!

I've noticed two things that should be addressed:
1. The sound file lists `buzzsaw_sawing` and `buzzsaw_grinding`, but the lang file entries are `buzzsaw_harvest_sawing` and `buzzsaw_harvest_grinding`
2. The sound events for attacking seem to not actually play?

Commit:
fixed attack sounds not playing
changed variables named "holder" to "noisyToolHolder" to prevent possible confusion with Holders

Commit:
highlight temporary variable with "TODO" so it's not left in before PR accept x)

Commit:
removed leftover commented-out import

Commit:
fixed mismatch between sounds.json subtitle string and lang file name

Commit:
changed attack sound duration to get rid of gap in sound

BluSunrize committed:
Merge branch '1.21.1' into DieselToolSounds

BluSunrize reviewed (commented):
LGTM! Thank you so much <3

BluSunrize reviewed (approved):
LGTM! Thank you so much <3
